### ISSUE TITLE
Bug: [no-unncessary-condition] Incorrectly detects possible values of type around await

### ISSUE THREAD

**Issue by mdouglass** [typescript-eslint/typescript-eslint#9055]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.ts&code=IYZwngdgxgBAZgV2gFwJYHsIwLbFRACgEoAuGABQCd1tUQBTAHgDd1UATAPhgG8AoGDAA29ZDGbAhCejAC8MZJWkCYwAO54xBUJFjE53foMESpM%2BXEkMVAXyLEVqODAKnpRXisFRMIdCIA6IXQAcwIAcjcZOgUlenCiWxh6IQZPYxgfCD9A4LDIyWkYGMtU%2BMTBGz4qvlYOHDxCRKA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2jNANgJZLyYBO0AxgPYC2dyK0KkANFAA4ECuA5iUwABeAE9O0NFXJFO8ALSTipAPRp4MqgrET5VABbUA1i0gBdEAF8QQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
async function main(): Promise<void> {
  let value = true
  await (async () => {
    value = false
  })()
  if (value) {
    console.log('value is true')
  } else {
    console.log('value is false')
  }
}

void main()
```

### ESLint Config

```javascript
{
  "extends": ["eslint:recommended", "plugin:@typescript-eslint/strict-type-checked"]
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Linting should report either this error or no error
```
  6:7   error  Unnecessary conditional, value is always falsy  @typescript-eslint/no-unnecessary-condition
```

### Actual Result

Linting reports this error:
```
  6:7   error  Unnecessary conditional, value is always truthy  @typescript-eslint/no-unnecessary-condition
```

### Additional Info

If you run this code, it will print `value is false` even though the linter tells you the conditional is unnecessary because `value is always truthy`.

The problem seems to be that its not seeing the change of `value` from `true` to `false` inside the immediately awaited async function. I would be fine with this reporting no linter warning or printing the correct warning (`value is always falsy`).

**Comment by mdouglass** [comment]
Note, if you change the immediately awaited function to a regular IIFE, then you get the proper linter warning (`value is always falsy`)
```
  ;(() => {
    value = false
  })()
```

**Comment by bradzacher** [comment]
The rule is powered by types.
The type of the variable is `true`.
Hence the rule reports.

To confirm this - hover the variable name in the [playground](https://www.typescriptlang.org/play/?#code/IYZwngdgxgBAZgV2gFwJYHsIwLbFRACgEoAuGABQCd1tUQBTAHgDd1UATAPhgG8AoGDAA29ZDGbAhCejAC8MZJWkCYwAO54xBUJFjE53foMESpM+XEkMVAXyLEVqODAKnpRXisEB6bzAB6APxeMFCYIOgiAHRC6ADmBADkbjJ0Ckr0iUS2MPRCDJ7GoeGR9DHxSSkwaZb5mdmCNnxNfKwcOHiE2U].

**Comment by mdouglass** [comment]
Are you saying that this is an underlying TypeScript bug? The value of the variable is clearly not true.

**Comment by Josh-Cena** [comment]
Please read the docs. https://typescript-eslint.io/rules/no-unnecessary-condition/#when-not-to-use-it

**Comment by mdouglass** [comment]
Fun, I had actually re-visited the docs, but glossed over the 2nd paragraph as our project is and always has been strongly typed.

**Comment by kirkwaiblinger** [comment]
(musing) it is kind of curious that TS treats a sync iife differently than an awaited async iife. I'm guessing it's just since an async iife need not be awaited (immediately or eventually), whereas a synchronous iife always executes... synchronously. So maybe they only felt comfortable special-casing the synchronous iife case, but didn't want to bother with the immediately awaited async case 🤷. 

But regardless, (as others have pointed to) all bets are basically off for the correctness of this rule once variables are modified in closures. I actually was surprised it was even right in the sync iife case.
